File: src/main/scala/com/teambr/bookshelf/common/container/BaseContainer.scala
package com.teambr.bookshelf.common.container

import com.teambr.bookshelf.common.container.slots.IPhantomSlot
import com.teambr.bookshelf.util.InventoryUtils
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.inventory.{ClickType, Container, IInventory, Slot}
import net.minecraft.item.ItemStack
import net.minecraftforge.items.{IItemHandler, SlotItemHandler}

/**
 * This file was created for com.teambr.bookshelf.Bookshelf
 *
 * com.teambr.bookshelf.Bookshelf is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Paul Davis pauljoda
 * @since August 04, 2015
 */
abstract class BaseContainer(val playerInventory: IInventory, val inventory: IItemHandler) extends Container {

    val inventorySize = getInventorySizeNotPlayer

    //Sometimes we have two inventories that aren't the players
    def getInventorySizeNotPlayer : Int = inventory.getSlots

    /**
      * Adds an inventory grid to the container
      *
      * @param xOffset X pixel offset
      * @param yOffset Y pixel offset
      * @param width How many wide
      */
    def addInventoryGrid(xOffset : Int, yOffset : Int, width : Int) : Unit = {
        val height = Math.ceil(inventorySize.toDouble / width).asInstanceOf[Int]
        var slotId = 0
        for(y <- 0 until height) {
            for(x <- 0 until width) {
                addSlotToContainer(new SlotItemHandler(inventory, slotId, xOffset + x * 18, yOffset + y * 18))
                slotId += 1
            }
        }
    }

    /**
      * Adds a line of slots
      *
      * @param xOffset X offset
      * @param yOffset Y offset
      * @param start start slot number
      * @param count how many slots
      */
    protected def addInventoryLine(xOffset: Int, yOffset: Int, start: Int, count: Int) {
        addInventoryLine(xOffset, yOffset, start, count, 0)
    }

    /**
      * Adds a line of inventory slots with a margin around them
      *
      * @param xOffset X Offset
      * @param yOffset Y Offset
      * @param start The start slot id
      * @param count The count of slots
      * @param margin How much to pad the slots
      */
    def addInventoryLine(xOffset: Int, yOffset: Int, start: Int, count: Int, margin: Int) : Unit = {
        var slotId = start
        for(x <- 0 until count) {
            addSlotToContainer(new SlotItemHandler(inventory, slotId, xOffset + x * (18 + margin), yOffset))
            slotId += 1
        }
    }

    /**
      * Adds the player offset with Y offset
      *
      * @param offsetY How far down
      */
    def addPlayerInventorySlots(offsetY: Int) {
        addPlayerInventorySlots(8, offsetY)
    }

    /**
      * Adds player inventory at location, includes space between normal and hotbar
      *
      * @param offsetX X offset
      * @param offsetY Y offset
      */
    def addPlayerInventorySlots (offsetX: Int, offsetY: Int) : Unit = {
        for(row <- 0 until 3) {
            for(column <- 0 until 9) {
                addSlotToContainer(new Slot(playerInventory,
                    column + row * 9 + 9,
                    offsetX + column * 18,
                    offsetY + row * 18))
            }
        }
        
        for(slot <- 0 until 9) {
            addSlotToContainer(new Slot(playerInventory, slot, offsetX + slot * 18, offsetY + 58))
        }
    }
    
    override def canInteractWith(entityPlayer: EntityPlayer): Boolean = true

    override def slotClick(slotNumber : Int, mouseButton : Int, modifier : ClickType, player : EntityPlayer) : ItemStack = {
        val slot = if (slotNumber < 0) null else this.inventorySlots.get(slotNumber)
        if(slot.isInstanceOf[IPhantomSlot])
             slotClickPhantom(slot, mouseButton, modifier, player)
        else
             super.slotClick(slotNumber, mouseButton, modifier, player)
    }

    def slotClickPhantom(slot : Slot, mouseButton : Int, modifier : ClickType, player : EntityPlayer) : ItemStack = {
        var stack : ItemStack = null

        if(mouseButton == 2) {
            if (slot.asInstanceOf[IPhantomSlot].canAdjust)
                slot.putStack(null)
        } else if(mouseButton == 0 || mouseButton == 1) {
            val playerInv = player.inventory
            slot.onSlotChanged()
            val stackSlot = slot.getStack
            val stackHeld = playerInv.getItemStack

            if(stackSlot != null)
                stack = stackSlot.copy

            if(stackSlot == null) {
                if(stackHeld != null && slot.isItemValid(stackHeld))
                    fillPhantomSlot(slot, stackHeld, mouseButton, modifier)
            } else if(stackHeld == null) {
                adjustPhantomSlot(slot, mouseButton, modifier)
                slot.onPickupFromSlot(player, playerInv.getItemStack)
            } else if(slot.isItemValid(stackHeld)) {
                if(InventoryUtils.canStacksMerge(stackSlot, stackHeld))
                    adjustPhantomSlot(slot, mouseButton, modifier)
                else
                    fillPhantomSlot(slot, stackHeld, mouseButton, modifier)
            }
        }
        stack
    }

    def adjustPhantomSlot(slot : Slot, mouseButton : Int, modifier : ClickType) : Unit = {
        if(!slot.asInstanceOf[IPhantomSlot].canAdjust)
            return

        val stackSlot = slot.getStack
        var stackSize : Int = 0
        if(modifier == ClickType.QUICK_MOVE) //TODO Unsure
            stackSize = if(mouseButton == 0) (stackSlot.stackSize + 1) / 2 else stackSlot.stackSize * 2
        else
            stackSize = if(mouseButton == 0) stackSlot.stackSize - 1 else stackSlot.stackSize + 1

        if(stackSize > slot.getSlotStackLimit)
            stackSize = slot.getSlotStackLimit

        stackSlot.stackSize = stackSize

        if(stackSlot.stackSize <= 0)
            slot.putStack(null)
    }

    def fillPhantomSlot(slot : Slot, stackHelf : ItemStack, mouseButton : Int, modifier : ClickType) : Unit = {
        if(!slot.asInstanceOf[IPhantomSlot].canAdjust)
            return

        var stackSize = if(mouseButton == 0) stackHelf.stackSize else 1
        if(stackSize > slot.getSlotStackLimit)
            stackSize  = slot.getSlotStackLimit

        val phantomStack = stackHelf.copy()
        phantomStack.stackSize = stackSize

        slot.putStack(phantomStack)
    }

    protected def mergeItemStackSafe(stackToMerge: ItemStack, start: Int, stop: Int, reverse: Boolean): Boolean = {
        var inventoryChanged: Boolean = false
        val delta: Int = if (reverse) -1 else 1
        val slots: java.util.List[Slot] = getSlots
        if (stackToMerge.isStackable) {
            var slotId: Int = if (reverse) stop - 1 else start
            while (stackToMerge.stackSize > 0 && ((!reverse && slotId < stop) || (reverse && slotId >= start))) {
                val slot: Slot = slots.get(slotId)
                if (slot.isItemValid(stackToMerge) && !slot.isInstanceOf[IPhantomSlot]) {
                    val stackInSlot: ItemStack = slot.getStack
                    if (InventoryUtils.tryMergeStacks(stackToMerge, stackInSlot)) {
                        slot.onSlotChanged()
                        inventoryChanged = true
                    }
                }
                slotId += delta
            }
        }
        if (stackToMerge.stackSize > 0) {
            var slotId: Int = if (reverse) stop - 1 else start
            while ((!reverse && slotId < stop) || (reverse && slotId >= start)) {
                val slot: Slot = slots.get(slotId)
                if (slot.isItemValid(stackToMerge) && !slot.isInstanceOf[IPhantomSlot]) {
                    val stackInSlot: ItemStack = slot.getStack
                    if (stackInSlot == null) {
                        slot.putStack(stackToMerge.copy)
                        slot.onSlotChanged()
                        stackToMerge.stackSize = 0
                        return true
                    }
                }
                slotId += delta
            }
        }
        inventoryChanged
    }

    override def transferStackInSlot(player: EntityPlayer, slotId: Int): ItemStack = {
        val slot: Slot = inventorySlots.get(slotId)
        if (slot != null && slot.getHasStack) {
            val itemToTransfer: ItemStack = slot.getStack
            val copy: ItemStack = itemToTransfer.copy
            if (slotId < inventorySize) {
                if (!mergeItemStackSafe(itemToTransfer, inventorySize, inventorySlots.size, reverse = true)) return null
            }
            else if (!mergeItemStackSafe(itemToTransfer, 0, inventorySize, reverse = false)) return null
            if (itemToTransfer.stackSize == 0) slot.putStack(null)
            else slot.onSlotChanged()
            if (itemToTransfer.stackSize != copy.stackSize) return copy
        }
        null
    }

    def getInventorySize: Int = inventorySize

    def getSlots : java.util.List[Slot] = inventorySlots
}

File: src/main/java/com/teambr/bookshelf/common/container/BaseContainer.java
package com.teambr.bookshelf.common.container;

import com.teambr.bookshelf.common.container.slots.IPhantomSlot;
import com.teambr.bookshelf.util.InventoryUtils;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.InventoryPlayer;
import net.minecraft.inventory.ClickType;
import net.minecraft.inventory.IInventory;
import net.minecraft.inventory.Slot;
import net.minecraft.item.ItemStack;
import net.minecraftforge.items.IItemHandler;
import net.minecraftforge.items.SlotItemHandler;

import javax.annotation.Nullable;
import java.util.List;

/**
 * This file was created for Bookshelf
 *
 * Bookshelf is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Paul Davis - pauljoda
 * @since 2/13/2017
 */
public abstract class BaseContainer extends ContainerGeneric {
    // Variables
    protected IInventory playerInventory;
    protected IItemHandler inventory;

    protected int inventorySize;

    /**
     * Creates the contianer object
     * @param playerInventory The players inventory
     * @param inventory The tile/object inventory
     */
    public BaseContainer(IInventory playerInventory, IItemHandler inventory) {
        this.playerInventory = playerInventory;
        this.inventory = inventory;

        inventorySize = inventory.getSlots();
    }

    /**
     * Get the size of the inventory that isn't the players
     * @return The inventory size that doesn't count the player inventory
     */
    public int getInventorySizeNotPlayer() {
        return inventorySize;
    }

    /**
     * Adds the player offset with Y offset
     *
     * @param offsetY How far down
     */
    protected void addPlayerInventorySlots(int offsetY) {
        addPlayerInventorySlots(8, offsetY);
    }

    /**
     * Adds player inventory at location, includes space between normal and hotbar
     *
     * @param offsetX X offset
     * @param offsetY Y offset
     */
    protected void addPlayerInventorySlots(int offsetX, int offsetY) {
        for(int row = 0; row < 3; row++) {
            for(int column = 0; column < 9; column++)
                addSlotToContainer(new Slot(playerInventory,
                        column + row * 9 + 9,
                        offsetX + column * 18,
                        offsetY + row * 18));
        }

        for(int slot = 0; slot < 9; slot++)
            addSlotToContainer(new Slot(playerInventory, slot, offsetX + slot * 18, offsetY + 58));
    }

    /**
     * Adds a line of slots
     *
     * @param xOffset X offset
     * @param yOffset Y offset
     * @param start start slot number
     * @param count how many slots
     */
    protected void addInventoryLine(int xOffset, int yOffset, int start, int count) {
        addInventoryLine(xOffset, yOffset, start, count, 0);
    }

    /**
     * Adds a line of inventory slots with a margin around them
     *
     * @param xOffset X Offset
     * @param yOffset Y Offset
     * @param start The start slot id
     * @param count The count of slots
     * @param margin How much to pad the slots
     */
    protected void addInventoryLine(int xOffset, int yOffset, int start, int count, int margin) {
        int slotID = start;
        for(int x = 0; x < count; x++) {
            addSlotToContainer(new SlotItemHandler(inventory, slotID, xOffset + x * (18 + margin), yOffset));
            slotID++;
        }
    }

    /**
     * Adds an inventory grid to the container
     *
     * @param xOffset X pixel offset
     * @param yOffset Y pixel offset
     * @param width How many wide
     * @param start The start slot id
     */
    protected void addInventoryGrid(int xOffset, int yOffset, int width, int start) {
        int height = (int) Math.ceil(inventorySize / width);
        int slotID = start;
        for(int y = 0; y < height; y++) {
            for(int x = 0; x < width; x++) {
                addSlotToContainer(new SlotItemHandler(inventory, slotID, xOffset + x * 18, yOffset + y * 18));
                slotID++;
            }
        }
    }

    /**
     * The logic for when a phantom slot is clicked
     * @param slot The slot
     * @param mouseButton The mouse button
     * @param modifier The modifier
     * @param player The player
     * @return The stack
     */
    private ItemStack slotClickPhantom(Slot slot, int mouseButton, ClickType modifier, EntityPlayer player) {
        ItemStack stack = null;

        if(mouseButton == 2) {
            if(((IPhantomSlot)slot).canAdjust())
                slot.putStack(null);
        } else if(mouseButton == 0 || mouseButton == 1) {
            InventoryPlayer playerInv = player.inventory;
            slot.onSlotChanged();
            ItemStack stackSlot = slot.getStack();
            ItemStack stackHeld = playerInv.getItemStack();

            if(stackSlot != null)
                stack = stackSlot.copy();

            if(stackSlot == null) {
                if(stackHeld != null && slot.isItemValid(stackHeld))
                    fillPhantomSlot(slot, stackHeld, mouseButton, modifier);
            } else if(stackHeld == null) {
                adjustPhantomSlot(slot, mouseButton, modifier);
                slot.onTake(player, playerInv.getItemStack());
            } else if(slot.isItemValid(stackHeld)) {
                if(InventoryUtils.canStacksMerge(stackSlot, stackHeld))
                    adjustPhantomSlot(slot, mouseButton, modifier);
                else
                    fillPhantomSlot(slot, stackHeld, mouseButton, modifier);
            }
        }

        return stack;
    }

    /**
     * Used to adjust the items in the phantom slot
     * @param slot The slot being clicked
     * @param mouseButton The mouse button
     * @param modifier The modifier
     */
    private void adjustPhantomSlot(Slot slot, int mouseButton, ClickType modifier) {
        if(!((IPhantomSlot)slot).canAdjust())
            return;

        ItemStack stackSlot = slot.getStack();
        int stackSize = 0;
        if(modifier == ClickType.QUICK_MOVE)
            stackSize = (mouseButton == 0) ? (stackSlot.getCount() + 1) / 2 : stackSlot.getCount() * 2;
        else
            stackSize =  (mouseButton == 0) ? stackSlot.getCount() - 1 : stackSlot.getCount() + 1;

        if(stackSize > slot.getSlotStackLimit())
            stackSize = slot.getSlotStackLimit();

        stackSlot.setCount(stackSize);

        if(stackSlot.getCount() <= 0)
            slot.putStack(null);
    }

    /**
     * Fills the phantom slot with the given slot, not consuming the held stack
     * @param slot The slot
     * @param stackHeld The stack held
     * @param mouseButton The mouse button
     * @param modifier The modifier
     */
    private void fillPhantomSlot(Slot slot, ItemStack stackHeld, int mouseButton, ClickType modifier) {
        if(!((IPhantomSlot)slot).canAdjust())
            return;

        int stackSize = (mouseButton == 0) ? stackHeld.getCount() : 1;

        if(stackSize > slot.getSlotStackLimit())
            stackSize = slot.getSlotStackLimit();

        ItemStack phantomStack = stackHeld.copy();
        phantomStack.setCount(stackSize);

        slot.putStack(phantomStack);
    }

    /*******************************************************************************************************************
     * Container                                                                                                       *
     *******************************************************************************************************************/

    /**
     * The logic for when a slot is clicked
     * @param slotId The slot
     * @param dragType The mouse button
     * @param clickTypeIn The modifier
     * @param player The player
     * @return The stack
     */
    @Nullable
    @Override
    public ItemStack slotClick(int slotId, int dragType, ClickType clickTypeIn, EntityPlayer player) {
        Slot slot = (slotId < 0) ? null : inventorySlots.get(slotId);
        if(slot != null) {
            if(slot instanceof IPhantomSlot)
                return slotClickPhantom(slot, dragType, clickTypeIn, player);
        }
        return super.slotClick(slotId, dragType, clickTypeIn, player);
    }

    /**
     * Take a stack from the specified inventory slot.
     */
    @Nullable
    @Override
    public ItemStack transferStackInSlot(EntityPlayer playerIn, int index) {
        if(index < 0 || index > inventorySlots.size())
            return super.transferStackInSlot(playerIn, index);
        Slot slot = inventorySlots.get(index);
        if(slot != null && slot.getHasStack()) {
            ItemStack itemToTransfer = slot.getStack();
            ItemStack copy = itemToTransfer.copy();

            if(index < getInventorySizeNotPlayer()) {
                if (!mergeItemStack(itemToTransfer, getInventorySizeNotPlayer(), inventorySlots.size(), true))
                    return null;
            } else if(!mergeItemStack(itemToTransfer, 0, getInventorySizeNotPlayer(), false))
                return null;

            if(itemToTransfer.getCount() == 0)
                slot.putStack(null);
            else
                slot.onSlotChanged();

            if(itemToTransfer.getCount() != copy.getCount())
                return copy;
        }
        return null;
    }
}
